crates/cargo_witgen/src/app.rs

```diff
@@ -1,79 +1,120 @@
 use anyhow::{bail, Context, Result};
-use clap::Parser;
+use clap::{Args, Parser, Subcommand};
 use std::{
     fs::{read, OpenOptions},
     io::Write,
     path::{Path, PathBuf},
 };
-use witgen_macro_helper::parse_crate_as_file;
+use witgen_macro_helper::{Wit, parse_crate_as_file};
+use syn::File;
 
 #[derive(Parser, Debug)]
-#[clap(author = "Benjamin Coenen <benjamin.coenen@hotmail.com>, Willem Wyndham <willem@ahalabs.dev>")]
+#[clap(
+    author = "Benjamin Coenen <benjamin.coenen@hotmail.com>, Willem Wyndham <willem@ahalabs.dev>"
+)]
 pub struct App {
     #[clap(subcommand)]
     pub command: Command,
 }
 
-#[derive(Parser, Debug)]
+#[derive(Debug, Subcommand)]
 pub enum Command {
     /// Generate wit files
     #[clap(alias = "gen")]
-    Generate {
-        /// Specify output file to generate wit definitions
-        #[clap(long, short = 'i', default_value = "./src/lib.rs")]
-        input: PathBuf,
-
-        /// Specify output file to generate wit definitions
-        #[clap(long, short = 'o', default_value = "index.wit")]
-        output: PathBuf,
-
-        /// Specify prefix file to copy into top of the generated wit file
-        #[clap(long, short = 'p')]
-        prefix_file: Vec<PathBuf>,
-
-        /// Specify prefix string to copy into top of the generated wit file
-        /// `--prefix-string 'use * from "string.wit"'`
-        #[clap(long, short = 's')]
-        prefix_string: Vec<String>,
-
-        /// Print results to stdout instead of to file
-        #[clap(long)]
-        stdout: bool,
-    },
+    Generate(Witgen),
 }
 
-impl App {
-    pub fn run(&self) -> Result<()> {
-        let Command::Generate {
-            input,
-            output,
-            prefix_file,
-            prefix_string,
-            stdout,
-        } = &self.command;
-        if !input.exists() {
-            bail!("input {:?} doesn't exist", input);
-        }
-        let wit = parse_crate_as_file(input)?;
+#[derive(Debug, Args)]
+pub struct Witgen {
+    /// Specify input file to generate wit definitions from
+    #[clap(long, short = 'i')]
+    pub input: Option<PathBuf>,
+
+    /// Specify input directory to generate wit definitions from
+    ///
+    ///
+    /// Will expect library: `<input-dir>/src/lib.rs`
+    #[clap(long, short = 'd', default_value = ".")]
+    pub input_dir: PathBuf,
+
+    /// Specify output file to generate wit definitions
+    #[clap(long, short = 'o', default_value = "index.wit")]
+    pub output: PathBuf,
+
+    /// Specify prefix file to copy into top of the generated wit file
+    #[clap(long, short = 'p')]
+    pub prefix_file: Vec<PathBuf>,
 
+    /// Specify prefix string to copy into top of the generated wit file
+    /// `--prefix-string 'use * from "string.wit"'`
+    #[clap(long, short = 's')]
+    pub prefix_string: Vec<String>,
+
+    /// Print results to stdout instead of to file
+    #[clap(long)]
+    pub stdout: bool,
+}
+
+impl Witgen {
+
+    pub fn read_input(&self) -> Result<File> {
```

Comment:
Just a suggestion but it could return a generic type which implements Read trait. It could avoid future breaking changes. Feel free to ignore if it doesn't not relevant

Reply:
I plan on continue down a path of breaking changes.  I'm the main power user of this so for now moving fast and breaking things is fine with me!  Plus I think most users will be using the CLI anyway.

Once we I add the `wit` AST, which I also plan on extending with more types than the current "official" version (I will also make a PR to their repo with my updated AST.  My plan is have a conversion to and from their version.  For example, I think that we should allow named enums.  Since it can be translated to their AST with the separate out structs.